0000-friends_in_high_kindednesses.md

```diff
@@ -0,0 +1,478 @@
+- Feature Name: associated_type_constructors
+- Start Date: 2016-04-29
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Allow type constructors to be associated with traits. This is an incremental
+step toward a more general feature commonly called "higher-kinded types," which
+is often ranked highly as a requested feature by Rust users. This specific
+feature (associated type constructors) resolves one of the most common use
+cases for higher-kindedness, is a relatively simple extension to the type
+system compared to other forms of higher-kinded polymorphism, and is forward
+compatible with more complex forms of higher-kinded polymorphism that may be
+introduced in the future.
+
+
+# Motivation
+[motivation]: #motivation
+
+Consider the following trait as a representative motivating example:
+
+```rust
+trait StreamingIterator {
```

Comment:
Despite using `StreamingIterator` as the motivating example, this RFC does not even attempt to solve the problem of how a streaming iterator could be used as an iterator.